src/mimefactory.rs

```diff
@@ -956,16 +920,74 @@ impl MimeFactory {
+                    if chat.typ != Chattype::Broadcast {
+                        for peerstate in peerstates.iter().filter_map(|(state, _)| state.as_ref()) {
+                            let Some(key) = peerstate.peek_key(verified) else {
+                                continue;
+                            };
+
+                            let fingerprint = key.dc_fingerprint().hex();
+                            let cmd = msg.param.get_cmd();
+                            let should_do_gossip = cmd == SystemMessage::MemberAddedToGroup
+                                || cmd == SystemMessage::SecurejoinMessage
+                                || multiple_recipients && {
+                                    let gossiped_timestamp: Option<i64> = context
+                                        .sql
+                                        .query_get_value(
+                                            "SELECT timestamp
+                                         FROM gossip_timestamp
+                                         WHERE chat_id=? AND fingerprint=?",
+                                            (chat.id, &fingerprint),
+                                        )
+                                        .await?;
+
+                                    // `gossip_period == 0` is a special case for testing,
+                                    // enabling gossip in every message.
+                                    gossip_period == 0
+                                        || gossiped_timestamp
+                                            .is_none_or(|ts| now >= ts + gossip_period)
+                                };
+
+                            if !should_do_gossip {
+                                continue;
+                            }
+
+                            let header = Aheader::new(
+                                peerstate.addr.clone(),
+                                key.clone(),
+                                // Autocrypt 1.1.0 specification says that
+                                // `prefer-encrypt` attribute SHOULD NOT be included.
+                                EncryptPreference::NoPreference,
+                            )
+                            .to_string();
+
+                            message = message.header(
+                                "Autocrypt-Gossip",
+                                mail_builder::headers::raw::Raw::new(header),
+                            );
+
+                            context
+                                .sql
+                                .execute(
+                                    "INSERT INTO gossip_timestamp (chat_id, fingerprint, timestamp)
+                                     VALUES                       (?, ?, ?)
+                                     ON CONFLICT                  (chat_id, fingerprint)
+                                     DO UPDATE SET timestamp=MAX(timestamp, excluded.timestamp)",
```

Comment:
I suggest just `excluded.timestamp`, otherwise if the clock is set back, we won't gossip indefinitely long. And change `|ts| now >= ts + gossip_period` above to `|ts| now >= ts + gossip_period || now < ts`